README.md
# Java Test Harness

This is a Java application used to test other Java classes.  It uses
Java "reflection" and it works like this:

1. It looks into the folder "/tests" to see if it can find Java .java files
2. It prints a list of these files
3. The user can select a file
4. This application will compile that file then execute the "testDB" method

### Benefits for instructors:

- the instructor of a course with a large number of students can
easily test the code of many students
- the students get exposure to some cool Java things they might not
otherwise see like reflection, plus they must be aware that their
source code must not just work in THEIR applications but in OTHER PEOPLE's
applications!

## Table of Contents

- [Installation](#installation)
- [Javadocs](#javadocs)
- [Authors](#authors)
- [License](#license)
- [CHANGELOG.md](CHANGELOG.md)


## Changelog

For a detailed list of changes and version history, please refer to the [CHANGELOG.md](CHANGELOG.md) file.

## Installation

1. No installation necessary.  Just be sure that new Java ".java" files to 
test are located in the "/tests" subdirectory.
2. 2 drivers are provided (SQLite, PostgreSQL) but these can easily be
replaced if desired.

## Javadocs

Javadocs are the built-in documentation system in Java, so that with one
command you can create beautiful HTML files that show your source code
documentation.

This application does NOT use Javadocs

## Authors

Your friendly, neighborhood Dr. Ken and his little green friend

## License

Never operate a motor vehicle without the proper license.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1] - 2023-05-02

### Added
- Added try-with-resources to make the code more readable

### Changed
- NA

### Deprecated
- NA

### Fixed
- NA

### Removed
- NA

### Security
- NA




## [1.0] - 2023-04-023

### Added
- Initial release

### Changed
- NA

### Deprecated
- NA

### Fixed
- NA

### Removed
- NA

### Security
- NA



tests/MyTest.java
// This is the bare minimum you need - but don't modify this file!
public class MyTest {
    public static void main(String[] args) {
        // Standalone test code here
        testDB("mytest_database", "mytest_username", "mytest_password");
    }

    public static void testDB(Object db, Object user, Object pass) {
        String database = (String) db;
        String username = (String) user;
        String password = (String) pass;
        // Test code here
        System.out.println("Testing database: " + database);
        System.out.println("Username: " + username);
        System.out.println("Password: " + password);
    }
}


tests/Test_Myname.java
import java.sql.*;

/******************************************************************************
 * You can use this file to experiment with JDBC. As long as it contains
 * the "testDB" method, then the instructor can automatically test
 * this file, e.g. in a midterm examination or final examination.
 * 
 * @author Dr. Ken and his little green friend
 * @version 1.0
 * @since 2023-04-29
 */
public class Test_Myname {

    // You have to change these for your own PostgreSQL
    // Feel free to change these before uploading for submission
    private static String myURL = "jdbc:postgresql://127.0.0.1:5431/postgres";
    private static String myUsername = "postgres";
    private static String myPassword = "kenpostgres";

    public static void main(String[] args) {
         testDB(
             myURL,       // DB connection URL
             myUsername,  // DB username
             myPassword); // DB password    
        }



/******************************************************************************
 * This is where you put your JDBC code. Feel free to call other methods.
 * This is the method that will be invoked from a different application
 * for testing or grading.
 * 
 * @author Dr. Ken and his little green friend
 * @version 1.0
 * @since 2023-04-29
 */
    public static void testDB(Object db_url, Object user, Object pass) {
        String url = (String) db_url;
        String username = (String) user;
        String password = (String) pass;
        System.out.println("*** WARNING - Don't submit with your credentials ***");
        System.out.println("Testing url: " + url);
        System.out.println("Username: " + username);
        System.out.println("Password: " + password);

        // Test to see if the SQLite driver is installed correctly
        System.out.println("\n--------------------");
        System.out.println("Executing test: testSQLiteDriver");
        testSQLiteDriver("jdbc:sqlite::memory");

        // Test to see if the SQLite driver is installed correctly
        System.out.println("\n--------------------");
        System.out.println("Executing test: testPostgresDriver");
        testPostgresSQLDriver( url, username, password);

        // Print out some metadata about the database
        System.out.println("\n--------------------");
        System.out.println("Executing test: testPostgresShowMetadata");
        testPostgresShowMetadata( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing scrolling cursor test");
        scrollingCursorTest( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing test1");
        test2( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing test2");
        test2( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing test3");
        test3( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing test4");
        test4( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing test5");
        test5( url, username, password);

    } // End of method


    
/******************************************************************************
 * This is a little method that uses the "in memory" version of SQLite
 * and prints out the time.
 * 
 * @author Dr. Ken and his little green friend
 * @version 1.0
 * @since 2023-04-29
 */
    public static void testSQLiteDriver(String url) {

        try {
            // Load the SQLite JDBC driver class explicitly
            // I put this code in here for you to get
            // some additional information in case Java
            // is not finding your driver file! :-)
            Class.forName("org.sqlite.JDBC");
            System.out.println("SQLite JDBC driver found!");
        } catch (ClassNotFoundException e) {
            System.out.println("SQLite JDBC driver not found");
            return;
        } // End of try

        try (            
            Connection conn = DriverManager.getConnection(url);
            Statement stmt = conn.createStatement(); 
            ResultSet rs = stmt.executeQuery("SELECT datetime('now')")
        )
        {
            // Display the current time
            if (rs.next()) {
                System.out.println("Current time: " + rs.getString(1));
            } else {
                System.out.println("Could not retrieve the current time.");
            }

        } catch (SQLException e) {
            System.out.println(e.getMessage());
        } // End of try

        } // End of method



/******************************************************************************
 * This is a little method that tests if the PostgreSQL driver is working
 * by printing out the time
 * 
 * @author Dr. Ken and his little green friend
 * @version 1.0
 * @since 2023-04-29
 */
public static void testPostgresSQLDriver( String url, String username, String password) {

    try {
        // Load the PostgreSQL JDBC driver class explicitly
        // I put this code in here for you to get
        // some additional information in case Java
        // is not finding your driver file! :-)
        Class.forName("org.postgresql.Driver");
        System.out.println("PostgreSQL JDBC driver found!");
    } catch (ClassNotFoundException e) {
        System.out.println("PostgreSQL JDBC driver not found");
        return;
    } // End of try

    try {
        // Connect to the database
        Connection conn = DriverManager.getConnection(url, username, password);

        // Create a statement
        Statement stmt = conn.createStatement();

        // Execute a SQL query to get the current timestamp
        ResultSet rs = stmt.executeQuery("SELECT CURRENT_TIMESTAMP");

        // Print out the result
        if (rs.next()) {
            Timestamp timestamp = rs.getTimestamp(1);
            System.out.println("Current time: " + timestamp);
        }

        // Close the result set, statement, and connection
        rs.close();
        stmt.close();
        conn.close();
        conn = null; // this is good practice

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



/******************************************************************************
 * This is a little method that tests if the PostgreSQL driver is working
 * by printing out the time
 * 
 * @author Dr. Ken and his little green friend
 * @version 1.0
 * @since 2023-04-29
 */
public static void testPostgresShowMetadata( String url, String username, String password) {

    Connection conn = null;
    try {
        // Connect to the database
        conn = DriverManager.getConnection(url, username, password);
        
        // Get metadata about the database
        DatabaseMetaData meta = conn.getMetaData();
        String dbName = meta.getDatabaseProductName();
        String dbVersion = "*** TO-DO: get the database version ***";
        String dbDriverName = meta.getDriverName();
        String dbDriverVersion = "*** TO-DO: get the driver version ***";
        String dbUsername = "*** TO-DO: get the database user name ***";
        System.out.println("DB product: " + dbName);
        System.out.println("DB version: " + dbVersion);
        System.out.println("Driver name: " + dbDriverName);
        System.out.println("Driver version: " + dbDriverVersion);
        System.out.println("Database user name: " + dbUsername);
        
    } catch (SQLException e) {
        System.out.println("Error connecting to the database: " + e.getMessage());
    } finally {
        // Close the connection when we're done
        try {
            if (conn != null) {
                conn.close();
            }
        } catch (SQLException e) {
            System.out.println("Error closing the database connection: " + e.getMessage());
        }
    } // End of try

} // End of method



 /******************************************************************************
 * This method is a nice test of how you can move a cursor in a record set
 * forwards and backwards!  Why do you want to do this?  Maybe you have a GUI
 * and you need data in both directions when the user scrolls up or down?
 * This method also creates the table and inserts some dummy data!
 * 
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void scrollingCursorTest( String url, String username, String password) {

//            // create the objects using a try-with-resources so no explicit "close" is needed
//            try (Connection conn = DriverManager.getConnection(url, username, password)) {
//                conn.setAutoCommit(false);
//
//                try (Statement stmt = conn.createStatement()) {
//
//                    // Create the table
//                    String createTableQuery = "CREATE TABLE IF NOT EXISTS months (id SERIAL PRIMARY KEY, month VARCHAR(50));";
//                    stmt.executeUpdate(createTableQuery);
//
//                    // Insert rows
//                    String[] monthsArray = {"January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"};
//                    for (String month : monthsArray) {
//                        String insertRowQuery = "INSERT INTO months (month) VALUES ('" + month + "');";
//                        stmt.executeUpdate(insertRowQuery);
//                    }
//
//                    // Commit the changes
//                    conn.commit();
//
//                } catch (SQLException ex) {
//                    conn.rollback();
//                    ex.printStackTrace();
//                }
//
//                // create the objects using a try-with-resources so no explicit "close" is needed
//                // In order to make the cursor scrollable (top-to-bottom and bottom-to-top)
//                // you need to use the resultset parameter TYPE_SCROLL_INSENSITIVE
//                try (Statement stmt = conn.createStatement(ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY)) {
//                    String query = "SELECT id, month FROM months";
//                    ResultSet resultSet = stmt.executeQuery(query);
//
//                    try (resultSet) {
//                        System.out.println("Forward:");
//                        while (resultSet.next()) {
//                            int id = resultSet.getInt("id");
//                            String month = resultSet.getString("month");
//                            System.out.println("ID: " + id + ", Month: " + month);
//                        }
//
//                        System.out.println("Backward:");
//                        while (resultSet.previous()) {
//                            int id = resultSet.getInt("id");
//                            String month = resultSet.getString("month");
//                            System.out.println("ID: " + id + ", Month: " + month);
//                        }
//                    }
//                } catch (SQLException ex) {
//                    ex.printStackTrace();
//                }
//
//            } catch (SQLException ex) {
//                ex.printStackTrace();
//            }


} // End of method


/******************************************************************************
 * You can use this for JDBC tests
 *   
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void test1( String url, String username, String password) {

    // try-with-resources
    try (Connection conn = DriverManager.getConnection(url, username, password);) 
    {        
        // add your code here

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



/******************************************************************************
 * You can use this for JDBC tests
 *   
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void test2( String url, String username, String password) {

    // try-with-resources
    try (Connection conn = DriverManager.getConnection(url, username, password);) 
    {        
        // add your code here

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



/******************************************************************************
 * You can use this for JDBC tests
 *   
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void test3( String url, String username, String password) {

    // try-with-resources
    try (Connection conn = DriverManager.getConnection(url, username, password);) 
    {        
        // add your code here

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



/******************************************************************************
 * You can use this for JDBC tests
 *   
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void test4( String url, String username, String password) {

    // try-with-resources
    try (Connection conn = DriverManager.getConnection(url, username, password);) 
    {        
        // add your code here

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



/******************************************************************************
 * You can use this for JDBC tests.  This example uses try-with-resources
 * e.g. try (...) {}. Within the parenthesis you can create your connection
 * object. Java will automatically guarantee that this will be closed when
 * you exit the try block, without you needing to explicity close the connection
 * e.g. Java will do this for you. It is a way to handle the connection well
 * but keep your code clean!
 *   
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void test5( String url, String username, String password) {

     // try-with-resources
     try (Connection conn = DriverManager.getConnection(url, username, password);) 
    {        
        // add your code here

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



} // End of class











tests/TestTemplate_Myname.java
import java.sql.*;

/******************************************************************************
 * You can use this file to experiment with JDBC. As long as it contains
 * the "testDB" method, then the instructor can automatically test
 * this file, e.g. in a midterm examination or final examination.
 * 
 * @author Dr. Ken and his little green friend
 * @version 1.0
 * @since 2023-04-29
 */
public class TestTemplate_Myname {

    // You have to change these for your own PostgreSQL
    private static String myURL = "jdbc:postgresql://127.0.0.1:5431/postgres";
    private static String myUsername = "postgres";
    private static String myPassword = "kenpostgres";

    public static void main(String[] args) {
         testDB(
             myURL,       // DB connection URL
             myUsername,  // DB username
             myPassword); // DB password    
        }



/******************************************************************************
 * This is where you put your JDBC code. Feel free to call other methods.
 * This is the method that will be invoked from a different application
 * for testing or grading.
 * 
 * @author Dr. Ken and his little green friend
 * @version 1.0
 * @since 2023-04-29
 */
    public static void testDB(Object db_url, Object user, Object pass) {
        String url = (String) db_url;
        String username = (String) user;
        String password = (String) pass;
        System.out.println("Testing url: " + url);
        System.out.println("Username: " + username);
        System.out.println("Password: " + password);

        // Test to see if the SQLite driver is installed correctly
        System.out.println("\n--------------------");
        System.out.println("Executing test: testSQLiteDriver");
        testSQLiteDriver("jdbc:sqlite::memory");

        // Test to see if the SQLite driver is installed correctly
        System.out.println("\n--------------------");
        System.out.println("Executing test: testPostgresDriver");
        testPostgresSQLDriver( url, username, password);

        // Print out some metadata about the database
        System.out.println("\n--------------------");
        System.out.println("Executing test: testPostgresShowMetadata");
        testPostgresShowMetadata( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing scrolling cursor test");
        scrollingCursorTest( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing test1");
        test2( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing test2");
        test2( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing test3");
        test3( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing test4");
        test4( url, username, password);

        // Execute the test
        System.out.println("\n--------------------");
        System.out.println("Executing test5");
        test5( url, username, password);

    } // End of method


    
/******************************************************************************
 * This is a little method that uses the "in memory" version of SQLite
 * and prints out the time.
 * 
 * @author Dr. Ken and his little green friend
 * @version 1.0
 * @since 2023-04-29
 */
    public static void testSQLiteDriver(String url) {

        try {
            // Load the SQLite JDBC driver class explicitly
            // I put this code in here for you to get
            // some additional information in case Java
            // is not finding your driver file! :-)
            Class.forName("org.sqlite.JDBC");
            System.out.println("SQLite JDBC driver found!");
        } catch (ClassNotFoundException e) {
            System.out.println("SQLite JDBC driver not found");
            return;
        } // End of try

        try (            
            Connection conn = DriverManager.getConnection(url);
            Statement stmt = conn.createStatement(); 
            ResultSet rs = stmt.executeQuery("SELECT datetime('now')")
        )
        {
            // Display the current time
            if (rs.next()) {
                System.out.println("Current time: " + rs.getString(1));
            } else {
                System.out.println("Could not retrieve the current time.");
            }

        } catch (SQLException e) {
            System.out.println(e.getMessage());
        } // End of try

        } // End of method



/******************************************************************************
 * This is a little method that tests if the PostgreSQL driver is working
 * by printing out the time
 * 
 * @author Dr. Ken and his little green friend
 * @version 1.0
 * @since 2023-04-29
 */
public static void testPostgresSQLDriver( String url, String username, String password) {

    try {
        // Load the PostgreSQL JDBC driver class explicitly
        // I put this code in here for you to get
        // some additional information in case Java
        // is not finding your driver file! :-)
        Class.forName("org.postgresql.Driver");
        System.out.println("PostgreSQL JDBC driver found!");
    } catch (ClassNotFoundException e) {
        System.out.println("PostgreSQL JDBC driver not found");
        return;
    } // End of try

    try {
        // Connect to the database
        Connection conn = DriverManager.getConnection(url, username, password);

        // Create a statement
        Statement stmt = conn.createStatement();

        // Execute a SQL query to get the current timestamp
        ResultSet rs = stmt.executeQuery("SELECT CURRENT_TIMESTAMP");

        // Print out the result
        if (rs.next()) {
            Timestamp timestamp = rs.getTimestamp(1);
            System.out.println("Current time: " + timestamp);
        }

        // Close the result set, statement, and connection
        rs.close();
        stmt.close();
        conn.close();
        conn = null; // this is good practice

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



/******************************************************************************
 * This is a little method that tests if the PostgreSQL driver is working
 * by printing out the time
 * 
 * @author Dr. Ken and his little green friend
 * @version 1.0
 * @since 2023-04-29
 */
public static void testPostgresShowMetadata( String url, String username, String password) {

    Connection conn = null;
    try {
        // Connect to the database
        conn = DriverManager.getConnection(url, username, password);
        
        // Get metadata about the database
        DatabaseMetaData meta = conn.getMetaData();
        String dbName = meta.getDatabaseProductName();
        String dbVersion = "*** TO-DO: get the database version ***";
        String dbDriverName = meta.getDriverName();
        String dbDriverVersion = "*** TO-DO: get the driver version ***";
        String dbUsername = "*** TO-DO: get the database user name ***";
        System.out.println("DB product: " + dbName);
        System.out.println("DB version: " + dbVersion);
        System.out.println("Driver name: " + dbDriverName);
        System.out.println("Driver version: " + dbDriverVersion);
        System.out.println("Database user name: " + dbUsername);
        
    } catch (SQLException e) {
        System.out.println("Error connecting to the database: " + e.getMessage());
    } finally {
        // Close the connection when we're done
        try {
            if (conn != null) {
                conn.close();
            }
        } catch (SQLException e) {
            System.out.println("Error closing the database connection: " + e.getMessage());
        }
    } // End of try

} // End of method



 /******************************************************************************
 * This method is a nice test of how you can move a cursor in a record set
 * forwards and backwards!  Why do you want to do this?  Maybe you have a GUI
 * and you need data in both directions when the user scrolls up or down?
 * This method also creates the table and inserts some dummy data!
 * 
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void scrollingCursorTest( String url, String username, String password) {

//            // create the objects using a try-with-resources so no explicit "close" is needed
//            try (Connection conn = DriverManager.getConnection(url, username, password)) {
//                conn.setAutoCommit(false);
//
//                try (Statement stmt = conn.createStatement()) {
//
//                    // Create the table
//                    String createTableQuery = "CREATE TABLE IF NOT EXISTS months (id SERIAL PRIMARY KEY, month VARCHAR(50));";
//                    stmt.executeUpdate(createTableQuery);
//
//                    // Insert rows
//                    String[] monthsArray = {"January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"};
//                    for (String month : monthsArray) {
//                        String insertRowQuery = "INSERT INTO months (month) VALUES ('" + month + "');";
//                        stmt.executeUpdate(insertRowQuery);
//                    }
//
//                    // Commit the changes
//                    conn.commit();
//
//                } catch (SQLException ex) {
//                    conn.rollback();
//                    ex.printStackTrace();
//                }
//
//                // create the objects using a try-with-resources so no explicit "close" is needed
//                // In order to make the cursor scrollable (top-to-bottom and bottom-to-top)
//                // you need to use the resultset parameter TYPE_SCROLL_INSENSITIVE
//                try (Statement stmt = conn.createStatement(ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY)) {
//                    String query = "SELECT id, month FROM months";
//                    ResultSet resultSet = stmt.executeQuery(query);
//
//                    try (resultSet) {
//                        System.out.println("Forward:");
//                        while (resultSet.next()) {
//                            int id = resultSet.getInt("id");
//                            String month = resultSet.getString("month");
//                            System.out.println("ID: " + id + ", Month: " + month);
//                        }
//
//                        System.out.println("Backward:");
//                        while (resultSet.previous()) {
//                            int id = resultSet.getInt("id");
//                            String month = resultSet.getString("month");
//                            System.out.println("ID: " + id + ", Month: " + month);
//                        }
//                    }
//                } catch (SQLException ex) {
//                    ex.printStackTrace();
//                }
//
//            } catch (SQLException ex) {
//                ex.printStackTrace();
//            }


} // End of method


/******************************************************************************
 * You can use this for JDBC tests
 *   
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void test1( String url, String username, String password) {

    // try-with-resources
    try (Connection conn = DriverManager.getConnection(url, username, password);) 
    {        
        // add your code here

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



/******************************************************************************
 * You can use this for JDBC tests
 *   
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void test2( String url, String username, String password) {

    // try-with-resources
    try (Connection conn = DriverManager.getConnection(url, username, password);) 
    {        
        // add your code here

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



/******************************************************************************
 * You can use this for JDBC tests
 *   
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void test3( String url, String username, String password) {

    // try-with-resources
    try (Connection conn = DriverManager.getConnection(url, username, password);) 
    {        
        // add your code here

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



/******************************************************************************
 * You can use this for JDBC tests
 *   
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void test4( String url, String username, String password) {

    // try-with-resources
    try (Connection conn = DriverManager.getConnection(url, username, password);) 
    {        
        // add your code here

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



/******************************************************************************
 * You can use this for JDBC tests.  This example uses try-with-resources
 * e.g. try (...) {}. Within the parenthesis you can create your connection
 * object. Java will automatically guarantee that this will be closed when
 * you exit the try block, without you needing to explicity close the connection
 * e.g. Java will do this for you. It is a way to handle the connection well
 * but keep your code clean!
 *   
 * @author (your name here)
 * @version 1.0
 * @since (your date here)
 */
public static void test5( String url, String username, String password) {

     // try-with-resources
     try (Connection conn = DriverManager.getConnection(url, username, password);) 
    {        
        // add your code here

    } catch (SQLException ex) {
        ex.printStackTrace();
    } // End of try
} // End of method



} // End of class











src/SQL.java
import java.io.File;
import java.io.FilenameFilter;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Scanner;

/**
 * This is a "test harness." The idea is that there will be a set of 
 * class files located in a directory /tests. This application will print
 * out that list, allow the user to pick one of those test files, it will then
 * compile that test file and execute it.
 *
 * @version 1.0
 * @since 2023-04-29
 * @author Dr. Ken
 */
 public class SQL {

    // You have to change these for your own PostgreSQL
    private static String myURL = "jdbc:postgresql://127.0.0.1:5431/postgres";
    private static String myUsername = "postgres";
    private static String myPassword = "kenpostgres";
    public static void main(String[] args) throws ClassNotFoundException, IllegalAccessException, InstantiationException, InvocationTargetException, NoSuchMethodException, SecurityException {
        File dir = new File("./tests/");
        File[] files = dir.listFiles(new FilenameFilter() {
            public boolean accept(File dir, String name) {
                return name.toLowerCase().endsWith(".java");
            }
        });
 
        Scanner scanner = new Scanner(System.in);
        while (true) {
            if (files != null && files.length > 0) {
                // List available test files
                System.out.println("\n\n----------");
                for (int i = 0; i < files.length; i++) {
                    System.out.println((i + 1) + ": " + files[i].getName());
                }
        
                // Prompt user for test file selection
                System.out.print("Enter the number of the test file to run (or 'q' to quit): ");
                String input = scanner.nextLine();
        
                if (input.equalsIgnoreCase("q")) {
                    // User wants to quit
                    scanner.close();
                    break;
                }
        
                int selection = Integer.parseInt(input);
        
                // Load and execute selected test file
                File selectedFile = files[selection - 1];
                String className = selectedFile.getName().replace(".java", "");
                Class<?> testClass = Class.forName(className);
                Method testMethod = testClass.getDeclaredMethod("testDB", Object.class, Object.class, Object.class);
                testMethod.invoke(null, myURL, myUsername, myPassword);
                // testMethod.invoke(null, "jdbc:postgresql://127.0.0.1:5431/postgres", "postgres", "kenpostgres");
            } else {
                System.err.println("No test files found in " + dir.getAbsolutePath());
                break;
            }
        }
        

    }
 }
 